design/src/rfcs/rfc0026_custom_validation_exceptions.md

```diff
@@ -0,0 +1,671 @@
+    // Translate the error messages to Spanish.
+    // ...
+}
+```
+
+```rust
+let my_service = MyService::builder_without_plugins()
+    .operation1(operation_handler1)
+    ...
+    .operationN(operation_handlerN)
+    .on_validation_failure(common_map_validation_failure_to_error)
+    .build()
+    .unwrap();
+```
+
+The framework would have to wrap the common `error::ValidationCustomError` into
+each of the `error::Operation1Error`, ... `error::OperationNError` enums under
+the hood.
+
+This approach has of course the obvious disadvantage that the application owner
+loses out in terms of:
+
+- customizability: they are no longer able to tweak the mapping for a specific
+  operation; and
+- type safety / expressiveness: they now have to work with essentially a vector
+  of strings as input.
+
+A combined solution where we let service owners register a common
+`on_validation_failure` callback while retaining the ability to customize the
+mapping for concrete operations:
+
+```rust
+let my_service = MyService::builder_without_plugins()
+    .operation1(operation_handler1)
+    .operation2(operation_handler2)
+    ...
+    .operationI(operation_handlerI)
+    .on_operationI_validation_failure(map_operationI_validation_failure_to_error)
+    ...
+    .operationN(operation_handlerN)
+    .on_validation_failure(common_map_validation_failure_to_error)
+    .build()
+    .unwrap();
+```
+
+is likely to lead to confusion.
+
+#### Registering the custom validation error mapping together with the operation
+
+The mapping from an operation's `ConstraintViolationExceptions` to the
+operation error shape is likely to live next to the operation handler
+implementation. Likewise, registering the mapping and the operation handler on
+the service builder is likely to occur in neighboring lines. Since the builder
+will fail if either of the two are not provided, we could couple the
+registration of both in a single call by passing in a pair:
+
+```rust
+let my_service = MyService::builder_without_plugins()
+    .my_operation((operation_handler, map_operation_validation_failure_to_error))
+    .build()
+    .unwrap();
+```
+
+This API [resembles more][smithy-typescript-custom-validation]
+smithy-typescript's:
+
+```typescript
+const handler = getGetForecastHandler(async (input) => getForecast(input),
+    (ctx, failures) => {
+        return new BadInputError(`${failures.length} bad inputs detected.`);
+    };
+});
+```
+
+#### Synchronicity of the custom validation error mapping
+
+Should the `ConstraintViolationExceptions` mapping be `async`? In other words,
+should we require a [`core::ops::Fn`], or a future type? All use cases so far
+would be solved with a synchronous mapping, but a priori there should be no
+problem in allowing async functions. As a data point, axum's
+[`IntoResponse`](https://docs.rs/axum-core/latest/axum_core/response/trait.IntoResponse.html)
+is synchronous.
```

Comment:
Would supporting async mappings be a backwards-compatible change?

Reply:
No, async functions essentially desugar to functions that return `Future`s.